File: synara/src/app/features/matrix-dto/media.ts
/**
 * Media handle DTO — metadata + handles only; no bytes, no key material.
 */

import type { MediaHandleId } from './ids';
import { hasForbiddenWireFields, isObject, optNumber, optString, reqString } from './parseUtil';

export const MEDIA_SOURCES = ['mxc', 'local_cache', 'upload'] as const;
export type MediaSource = typeof MEDIA_SOURCES[number];
const MEDIA_SOURCE_SET = new Set<string>(MEDIA_SOURCES);

export function isMediaSource(value: unknown): value is MediaSource {
  return typeof value === 'string' && MEDIA_SOURCE_SET.has(value);
}

export type MediaHandle = {
  handleId: MediaHandleId;
  mxcUri?: string;
  source?: MediaSource;
  mimeType?: string;
  sizeBytes?: number;
  width?: number;
  height?: number;
  durationMs?: number;
  thumbnailHandleId?: MediaHandleId;
};

export function parseMediaHandle(value: unknown): MediaHandle | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const handleId = reqString(value, 'handleId');
  const mxcUri = optString(value, 'mxcUri');
  const mimeType = optString(value, 'mimeType');
  const sizeBytes = optNumber(value, 'sizeBytes');
  const width = optNumber(value, 'width');
  const height = optNumber(value, 'height');
  const durationMs = optNumber(value, 'durationMs');
  const thumbnailHandleId = optString(value, 'thumbnailHandleId');
  if (
    handleId === null ||
    mxcUri === null ||
    mimeType === null ||
    sizeBytes === null ||
    width === null ||
    height === null ||
    durationMs === null ||
    thumbnailHandleId === null
  ) {
    return null;
  }
  let source: MediaSource | undefined;
  if (value.source !== undefined) {
    if (!isMediaSource(value.source)) return null;
    source = value.source;
  }
  return {
    handleId,
    mxcUri,
    source,
    mimeType,
    sizeBytes,
    width,
    height,
    durationMs,
    thumbnailHandleId,
  };
}

File: crates/synara-core/src/dto/media.rs
//! Media handle DTO — metadata + handles only; **no bytes**, no key material.

use serde::{Deserialize, Serialize};

use super::ids::MediaHandleId;

/// Where the media content is sourced from (product enum).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MediaSource {
    Mxc,
    LocalCache,
    Upload,
}

impl MediaSource {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Mxc => "mxc",
            Self::LocalCache => "local_cache",
            Self::Upload => "upload",
        }
    }
}

/// Product media handle — references content without embedding bytes.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaHandle {
    pub handle_id: MediaHandleId,
    /// Optional mxc URI string (never resolved bytes on wire).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mxc_uri: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<MediaSource>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub width: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub height: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub thumbnail_handle_id: Option<MediaHandleId>,
}
